fendermint/vm/topdown/src/finality/null.rs

```diff
@@ -180,18 +183,41 @@ impl FinalityWithNull {
         } else {
             unreachable!("last committed finality will be available at this point");
         };
-        let next_proposal_height = last_committed_height + self.min_proposal_interval;
 
-        if next_proposal_height > latest_height {
-            tracing::debug!("proposal period not reached yet");
-            return Ok(None);
+        let max_proposal_height = last_committed_height + self.config.max_proposal_range();
+        let candidate_height = min(max_proposal_height, latest_height);
+        tracing::debug!(max_proposal_height, candidate_height, "propose heights");
+
+        let first_non_null_height =
+            if let Some(h) = self.first_non_null_block_before(candidate_height)? {
+                h
+            } else {
+                tracing::debug!(height = candidate_height, "no non-null block found before");
+                return Ok(None);
+            };
+
+        tracing::debug!(first_non_null_height, candidate_height);
+        // an extra layer of delay
+        let r =
+            self.first_non_null_block_before(first_non_null_height - self.config.proposal_delay())?;
+        tracing::debug!(delayed_height = r, delay = self.config.proposal_delay());
+        if let Some(final_proposal) = r {
+            // this is possible due to delayed execution as the proposed height's data cannot be
+            // executed because they have yet to be executed.
+            return if last_committed_height == final_proposal {
```

Comment:
@adlrocha Please help review this check, it is possible under delayed execution.